File: mine/tas/src/main/java/org/ta/trader/speculate/TaBaseOrderOpenner.java
package org.ta.trader.speculate;


import org.ta.trader.TaConstants;
import org.ta.trader.TaSide;
import org.ta.trader.TaSymbol;
import org.ta.trader.executing.TaOrderOpeningAction;
import org.ta.trader.hista.syriod.TaSyriodWindow;
import org.ta.trader.manage.TaUserSession;
import org.ta.trader.rating.TaRateMonitor;

public abstract class TaBaseOrderOpenner implements TaOrderOpenner {

	protected TaSide direction;

	protected TaSyriodWindow window;
	
	protected double maxOrderRisk;
	
	protected double minOrderRisk;
	
	protected double takeProfitRate = 1.0;	
	
	protected String advisorId;
	
	protected TaRateMonitor quotesMonitor;
	
	protected TaSymbol symbol;
	
	public TaBaseOrderOpenner(String aid, TaSide dir, TaSyriodWindow window) {
		super();
		this.advisorId = aid;
		this.window = window;
		this.direction = dir;
		this.maxOrderRisk = TaConstants.OST_OPEN_ORDER_MAX_RISK;
		this.minOrderRisk = TaConstants.OST_OPEN_ORDER_MIN_RISK;
		this.symbol = window.getSyriod().getSymbol();
	}

	public void setTakeProfitRate(double pr) {
		this.takeProfitRate = pr;
	}

	protected double orderMarketPrice(TaSide dir) {
		if (dir.isUp()) {
			return this.quotesMonitor.getRate(this.symbol).getAsk();
		} else {
			return this.quotesMonitor.getRate(this.symbol).getBid();
		}
	}
	protected double orderStoploss(TaOrderOpeningAction oA) {
		double risk = oA.getInternalOrder().getRisk();
		double price = oA.getInternalOrder().getPrice();
		if (this.direction.isUp()) {
			risk = -risk;
		}
		return price + risk;
	}
	
	double orderTakeprofit(TaOrderOpeningAction oA) {
		double risk = oA.getInternalOrder().getRisk();
		double price = oA.getInternalOrder().getPrice();
		if (this.direction.isUp()) {
			return price + risk * this.takeProfitRate;
		} else {
			return price - risk * this.takeProfitRate;
		}
	}
	
	int exclusiveOrderType() {
		return -1;
	}
	
	protected boolean openOrder(TaOrderOpeningAction oA) {
		//this.doLog("sendOrder()");
		double risk =oA.getInternalOrder().getRisk();
		if (oA.getInternalOrder().getRisk() > this.maxOrderRisk) {
			//this.doLog(
			//		STRINGConcatenate("orderRisk:",
			//				TaLang::DoubleTo(oA.getInternalOrder().getRisk()),
			//				" is exceed the maxOrderRisk:",
			//				TaLang::DoubleTo(this.maxOrderRisk)));
			return false;
		}
		if (oA.getInternalOrder().getRisk() < this.minOrderRisk) {
			//this.doLog(
			//		STRINGConcatenate("orderRisk:",
			//				TaLang::DoubleTo(oA.getInternalOrder().getRisk()),
			//				" is lower the minOrderRisk:",
			//				TaLang::DoubleTo(this.minOrderRisk)));
			return false;
		}
		
		double stoploss = oA.getInternalOrder().getSl();

		if (stoploss == 0) {
			stoploss = this.orderStoploss(oA);
			oA.getInternalOrder().setSl(stoploss);
		}

		double takeprofit = oA.getInternalOrder().getTp();

		if (takeprofit == 0) {
			takeprofit = this.orderTakeprofit(oA);
			oA.getInternalOrder().setTp(takeprofit);
		}

		oA.getInternalOrder().setAdvisorId(this.advisorId);//
		oA.setOpenOrder(true);//
		return true;
	}

}

File: mine/mineea/src/main/cpp/ta/core/advisor/support/TaBaseOrderOpenner.cpp
#include "TaBaseOrderOpenner.h"


TaBaseOrderOpenner::TaBaseOrderOpenner(int pStrategy,STRING pName, const TaVDirRef& dir) :
		TaOrderOpenner(pStrategy,pName) {

	this->takeProfitRate = 1.0;
	this->direction = dir;
	this->maxOrderRisk = TaConstants::OST_OPEN_ORDER_MAX_RISK;
	this->minOrderRisk = TaConstants::OST_OPEN_ORDER_MIN_RISK;

}

TaBaseOrderOpenner::~TaBaseOrderOpenner() {

}
void TaBaseOrderOpenner::setContainer(TaContainerRef& cref){
	this->ref_orderSender =	cref.o()->getComponent(("componentOrderSender"));
	this->orderSender = this->ref_orderSender.obj();

}

void TaBaseOrderOpenner::setTakeProfitRate(double pRate){
	this->takeProfitRate = pRate;
}
double TaBaseOrderOpenner::orderStoploss(TaOrderOpeningAction* oA) {
	double risk = oA->getOrderRisk();
	double price = oA->getOrderPrice();
	if (this->direction.o()->isUp()) {
		risk = -risk;
	}
	return price + risk;
}

double TaBaseOrderOpenner::orderTakeprofit(TaOrderOpeningAction* oA) {
	double risk = oA->getOrderRisk();
	double price = oA->getOrderPrice();
	if (this->direction.o()->isUp()) {
		return price + risk * this->takeProfitRate;
	} else {
		return price - risk * this->takeProfitRate;
	}
}

double TaBaseOrderOpenner::orderMarketPrice(const TaVDirRef& dir) {
	if (dir.o()->isUp()) {
		return TaLang::getAsk();
	} else {
		return TaLang::getBid();
	}
}

int TaBaseOrderOpenner::exclusiveOrderType() {
	return -1;
}

bool TaBaseOrderOpenner::tryOpenOrder(TaOrderOpeningActionRef& oA) {
	return false;
}

void TaBaseOrderOpenner::sendOrder(TaOrderOpeningAction* oA) {
	//this->doLog("sendOrder()");
	double risk =oA->getOrderRisk();
	if (oA->getOrderRisk() > this->maxOrderRisk) {
		//this->doLog(
		//		STRINGConcatenate("orderRisk:",
		//				TaLang::DoubleTo(oA->getOrderRisk()),
		//				" is exceed the maxOrderRisk:",
		//				TaLang::DoubleTo(this->maxOrderRisk)));
		return;
	}
	if (oA->getOrderRisk() < this->minOrderRisk) {
		//this->doLog(
		//		STRINGConcatenate("orderRisk:",
		//				TaLang::DoubleTo(oA->getOrderRisk()),
		//				" is lower the minOrderRisk:",
		//				TaLang::DoubleTo(this->minOrderRisk)));
		return;
	}
	STRING symbol = TaLang::__Symbol();
	int cmd = oA->getOrderCmd();
	double lots = oA->getOrderLots();
	double price = oA->getOrderPrice();
	double stoploss = oA->getOrderStoploss();

	if (stoploss == 0) {
		stoploss = this->orderStoploss(oA);
		oA->setOrderStoploss(stoploss);
	}

	double takeprofit = oA->getOrderTakeprofit();

	if (takeprofit == 0) {
		takeprofit = this->orderTakeprofit(oA);
		oA->setOrderTakeprofit(takeprofit);
	}


	int oppositeOrderType = this->exclusiveOrderType();



	TaOrderComments* ocs = oA->getOrderComments();
	ocs->setOriginalRisk(risk);
	STRING comments = ocs->to();

	if (this->isContainsOrder(symbol, oppositeOrderType)) {
		//this->doLog(
		//		STRINGConcatenate("cannot create opposite order,symbol:",
		//				symbol, ",cmd:", cmd));
		return;
	}
	oA->setOrderMagic(this->strategy);
	this->orderSender->trySendOrder(oA);
}

bool TaBaseOrderOpenner::isContainsOrder(STRING pSymbol, int pOrderType) {
	int total = TaLang::__OrdersTotal();
	for (int i = 0; i < total; i++) {
		if (TaLang::__OrderSelect(i, TaLang::SELECT_BY_POS_, TaLang::MODE_TRADES_) == false) {
			continue;
		}
		if (TaLang::__OrderSymbol() != pSymbol) {
			continue;
		}
		if (TaLang::__OrderType() == pOrderType) {
			return true;
		}
	}
	return false;

}
